add node mapping processing

yushangdi opened the pull request:
Summary:
Add `node_mapping = create_node_mapping(pre_grad_graph_id, inductor_post_to_pre_grad_nodes, debug_info)`, to produce a `inductor_provenance_tracking_node_mappings.json` file. This file will be used by the provenance tracking highlighter tool to create provenance visualization.

`inductor_triton_kernel_to_post_grad_nodes.json` and `inductor_provenance_tracking_node_mappings.json` files are not dumped if they are both empty. So it's removed from some of the `test_structured_trace` tests. 

Test Plan:
CI
```
buck run mode/dev-nosan  fbcode//caffe2/test:fx -- -r graph_provenance

buck run mode/dev-nosan fbcode//caffe2/test/inductor:provenance_tracing

python test/dynamo/test_structured_trace.py
```

Differential Revision: D68190173




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

chenyang78 commented on torch/_inductor/debug.py:
```diff
@@ -656,6 +676,75 @@ class TensorMetadataHolder:
+    - pre_grad graph nodes and post_grad graph code nodes, and vice versa
+    - triton kernel name and post_grad graph code nodes, and vice versa
+    """
+    try:
```
Wondering under which condition we would hit exceptions. Could we narrow the scope of the ```try``` block so that we could get more accurate exception information like which statement caused the exceptiont. Thanks.

yushangdi replied:
@chenyang78 It would happen if the input json format is somehow not correct. Working on an improved version.

yushangdi replied:
@chenyang78 Added some additional logging msg! It should never reach the exception clause now, but I still left the exception clause just in case. Since this is just logging, it should never cause the main program to crash.

Skylion007 commented on torch/_inductor/debug.py:
```diff
@@ -656,6 +676,124 @@ class TensorMetadataHolder:
+    except Exception as e:
+        # Since this is just logging code, it should never interfere with regular
+        # program execution, so we use this try-except to guard against any error
+        log.error("Unexpected error in create_node_mapping: %s", str(e))
```
```suggestion
        log.error("Unexpected error in create_node_mapping: %s", e)
```
cast should be implicit with the '%' formatting

yushangdi replied:
Fixed now internally and export job in queue! Thanks!

yushangdi committed:
[provenance tracking] Add node mapping processing (#146103)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/146103

Add `node_mapping = create_node_mapping(pre_grad_graph_id, inductor_post_to_pre_grad_nodes, debug_info)`, to produce a `inductor_provenance_tracking_node_mappings.json` file. This stores all the node mapping information that the front-end provenance tracking highlighter tool needs.

`inductor_triton_kernel_to_post_grad_nodes` and `inductor_provenance_tracking_node_mappings` json files are not dumped if they are both empty.

Test Plan:
CI
```
buck run mode/dev-nosan  fbcode//caffe2/test:fx -- -r graph_provenance

buck run mode/dev-nosan fbcode//caffe2/test/inductor:provenance_tracing

python test/dynamo/test_structured_trace.py
```

Reviewed By: chenyang78, YUNQIUGUO

Differential Revision: D68190173
